File: generated_code/cpp/test_unittest_2903.cpp
#include <gtest/gtest.h>
#include <vector>
#include "solution_proxy.h"

// Helper function to create a linked list from a vector
ListNode* createList(const std::vector<int>& values) {
    if (values.empty()) return nullptr;
    ListNode* head = new ListNode(values[0]);
    ListNode* current = head;
    for (size_t i = 1; i < values.size(); ++i) {
        current->next = new ListNode(values[i]);
        current = current->next;
    }
    return head;
}

// Helper function to convert a linked list back to a vector
std::vector<int> listToVector(ListNode* head) {
    std::vector<int> result;
    ListNode* current = head;
    while (current != nullptr) {
        result.push_back(current->val);
        current = current->next;
    }
    return result;
}

// Helper function to safely delete a linked list to prevent memory leaks
void freeList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

class InsertGCDTest : public ::testing::Test {
protected:
    Solution solution;
};

// Test Example 1: Multiple elements with varying GCDs
TEST_F(InsertGCDTest, Example1_MultipleElements) {
    std::vector<int> inputValues = {18, 6, 10, 3};
    std::vector<int> expectedValues = {18, 6, 6, 2, 10, 1, 3};
    
    ListNode* head = createList(inputValues);
    ListNode* resultHead = solution.insertGreatestCommonDivisors(head);
    std::vector<int> resultValues = listToVector(resultHead);
    
    EXPECT_EQ(resultValues, expectedValues);
    freeList(resultHead);
}

// Test Example 2: Single element (Minimum node constraint)
TEST_F(InsertGCDTest, Example2_SingleElement) {
    std::vector<int> inputValues = {7};
    std::vector<int> expectedValues = {7};
    
    ListNode* head = createList(inputValues);
    ListNode* resultHead = solution.insertGreatestCommonDivisors(head);
    std::vector<int> resultValues = listToVector(resultHead);
    
    EXPECT_EQ(resultValues, expectedValues);
    freeList(resultHead);
}

// Test Edge Case: Two elements with a GCD of 1 (Coprime numbers)
TEST_F(InsertGCDTest, TwoElementsCoprime) {
    std::vector<int> inputValues = {13, 7};
    std::vector<int> expectedValues = {13, 1, 7};
    
    ListNode* head = createList(inputValues);
    ListNode* resultHead = solution.insertGreatestCommonDivisors(head);
    std::vector<int> resultValues = listToVector(resultHead);
    
    EXPECT_EQ(resultValues, expectedValues);
    freeList(resultHead);
}

// Test Edge Case: Two elements where one divides the other completely
TEST_F(InsertGCDTest, TwoElementsMultiples) {
    std::vector<int> inputValues = {5, 25};
    std::vector<int> expectedValues = {5, 5, 25};
    
    ListNode* head = createList(inputValues);
    ListNode* resultHead = solution.insertGreatestCommonDivisors(head);
    std::vector<int> resultValues = listToVector(resultHead);
    
    EXPECT_EQ(resultValues, expectedValues);
    freeList(resultHead);
}

// Test Edge Case: Identical elements (GCD is the number itself)
TEST_F(InsertGCDTest, IdenticalElements) {
    std::vector<int> inputValues = {12, 12, 12};
    std::vector<int> expectedValues = {12, 12, 12, 12, 12};
    
    ListNode* head = createList(inputValues);
    ListNode* resultHead = solution.insertGreatestCommonDivisors(head);
    std::vector<int> resultValues = listToVector(resultHead);
    
    EXPECT_EQ(resultValues, expectedValues);
    freeList(resultHead);
}

// Test Edge Case: Boundary node values (1 and 1000 based on constraints)
TEST_F(InsertGCDTest, BoundaryNodeValues) {
    std::vector<int> inputValues = {1, 1000};
    std::vector<int> expectedValues = {1, 1, 1000};
    
    ListNode* head = createList(inputValues);
    ListNode* resultHead = solution.insertGreatestCommonDivisors(head);
    std::vector<int> resultValues = listToVector(resultHead);
    
    EXPECT_EQ(resultValues, expectedValues);
    freeList(resultHead);
}

// Test Case: Maximum constraint value pair (1000 and 1000)
TEST_F(InsertGCDTest, MaxConstraintValues) {
    std::vector<int> inputValues = {1000, 1000};
    std::vector<int> expectedValues = {1000, 1000, 1000};
    
    ListNode* head = createList(inputValues);
    ListNode* resultHead = solution.insertGreatestCommonDivisors(head);
    std::vector<int> resultValues = listToVector(resultHead);
    
    EXPECT_EQ(resultValues, expectedValues);
    freeList(resultHead);
}

File: generated_code/python/test_unittest_2903.py
import os
import importlib.util
import unittest

# Dynamic loading of the solution module as per guidelines
solution_path = os.environ.get("TEST_SOLUTION_FILE")
if not solution_path:
    raise RuntimeError("TEST_SOLUTION_FILE environment variable is not set.")

spec = importlib.util.spec_from_file_location("Solution", solution_path)
sol_module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(sol_module)
Solution = sol_module.Solution
ListNode = sol_module.ListNode


class TestInsertGreatestCommonDivisors(unittest.TestCase):

    def _create_linked_list(self, arr):
        """Helper method to construct a linked list from an array."""
        if not arr:
            return None
        head = ListNode(arr[0])
        curr = head
        for val in arr[1:]:
            curr.next = ListNode(val)
            curr = curr.next
        return head

    def _linked_list_to_list(self, head):
        """Helper method to convert a linked list to a Python list for easy assertion."""
        result = []
        curr = head
        while curr:
            result.append(curr.val)
            curr = curr.next
        return result

    def test_example_1_multiple_nodes(self):
        """Test with multiple nodes having various GCD values (Example 1)."""
        input_list = [18, 6, 10, 3]
        expected_output = [18, 6, 6, 2, 10, 1, 3]
        
        head = self._create_linked_list(input_list)
        sol = Solution()
        result_head = sol.insertGreatestCommonDivisors(head)
        
        self.assertEqual(self._linked_list_to_list(result_head), expected_output)

    def test_example_2_single_node(self):
        """Test with a single node where no pairs exist (Example 2 / Minimum constraint)."""
        input_list = [7]
        expected_output = [7]
        
        head = self._create_linked_list(input_list)
        sol = Solution()
        result_head = sol.insertGreatestCommonDivisors(head)
        
        self.assertEqual(self._linked_list_to_list(result_head), expected_output)

    def test_two_nodes_same_value(self):
        """Test with two nodes having identical values (GCD should equal the value)."""
        input_list = [5, 5]
        expected_output = [5, 5, 5]
        
        head = self._create_linked_list(input_list)
        sol = Solution()
        result_head = sol.insertGreatestCommonDivisors(head)
        
        self.assertEqual(self._linked_list_to_list(result_head), expected_output)

    def test_two_nodes_coprime(self):
        """Test with two coprime nodes (GCD should be 1)."""
        input_list = [13, 7]
        expected_output = [13, 1, 7]
        
        head = self._create_linked_list(input_list)
        sol = Solution()
        result_head = sol.insertGreatestCommonDivisors(head)
        
        self.assertEqual(self._linked_list_to_list(result_head), expected_output)

    def test_maximum_constraint_values(self):
        """Test with maximum allowed node values (1000) to ensure correctness."""
        input_list = [1000, 500, 1000]
        expected_output = [1000, 500, 500, 500, 1000]
        
        head = self._create_linked_list(input_list)
        sol = Solution()
        result_head = sol.insertGreatestCommonDivisors(head)
        
        self.assertEqual(self._linked_list_to_list(result_head), expected_output)


if __name__ == "__main__":
    unittest.main()